src/main/java/com/proximyst/ban/service/impl/ImplMessageService.java

```diff
@@ -148,125 +118,140 @@ public ImplMessageService(final @NonNull IUserService userService,
+                "reason", punishment.getReason()
+                    .map(MiniMessage.get()::escapeTokens)
+                    .orElse("No reason specified"),
+                "punishmentType", punishment.getPunishmentType().name(),
+                "punishmentVerb", punishment.getPunishmentType().getVerbPastTense().map(this.cfg),
+
+                "expiry", !punishment.currentlyApplies()
+                    ? this.cfg.formatting.isLifted
+                    : punishment.isPermanent()
+                        ? this.cfg.formatting.never
+                        : DurationFormatUtils
+                            .formatDurationHMS(punishment.getExpiration() - System.currentTimeMillis()),
+                "duration", punishment.isPermanent()
+                    ? this.cfg.formatting.permanently
+                    : this.cfg.formatting.durationFormat
+                        .replace("<duration>",
+                            DurationFormatUtils.formatDurationWords(
+                                punishment.getExpiration() - System.currentTimeMillis(),
+                                false,
+                                false
+                            )))
+        )
+        .thenCompose(p -> this.formatMessage(messageKey, p));
   }
 
-  private @NonNull Component parseWithTarget(
-      final @NonNull String msg,
-      final @NonNull BanUser target) {
-    return MiniMessage.get().parse(
-        msg,
+  private @NonNull CompletableFuture<@NonNull Component> formatMessageFuture(final @NonNull String message,
+      final @Positive int futureCount, final @NonNull Object @NonNull ... placeholders) {
+    final CompletableFuture<?>[] futures = new CompletableFuture<?>[futureCount];
 
-        "targetName", target.getUsername(),
-        "targetUuid", target.getUuid().toString()
-    );
+    // We know how many futures there are, but we also need to actually find them.
+    int futureIdx = 0;
+    for (final Object placeholder : placeholders) {
+      if (placeholder instanceof CompletableFuture) {
+        futures[futureIdx++] = (CompletableFuture<?>) placeholder;
+      }
+    }
+
+    // We now need to know when all the futures are done.
+    // Luckily, we can do that with helper methods in CompletableFuture.
+    final CompletableFuture<?> allDone = CompletableFuture.allOf(futures);
+
+    return allDone.thenApply($ -> {
+      // We now know all the futures are done.
+      // We'll need to unwrap them, then pass to MiniMessage.
+      int found = 0;
+      for (int i = 0; i < placeholders.length; ++i) {
+        final Object obj = placeholders[i];
+        if (obj instanceof CompletableFuture) {
+          ++found;
+          placeholders[i] = String.valueOf(((CompletableFuture<?>) obj).join());
+        }
+
+        if (found == futureCount) {
+          break;
+        }
+      }
+
+      return MiniMessage.get().parse(message, (String[]) placeholders);
```

Comment:
Stop it :(

Reply:
This is not unsafe.

Reply:
Wouldn't it be better to add the strings to a new array with the correct type? That way you can prevent the cast all together.

Reply:
Also, it looks like this cast is the only actual unsafe cast out of the ones I said were unsafe. There is no guarantee that all items in the array are Strings, because you only convert the Futures into Strings. If someone were to input an array that contained some other type (let's say a Double), that element would remain in the array unchanged and would cause an exception here.

Reply:
It wouldn't be. It converts all non-futures to strings in the above loop.

Reply:
As far as I can tell, the `String#valueOf` call happens in a different method. Within the `formatMessageFuture` method, there is no guarantee that all elements of the `placeholders` array are actually Strings or Futures.

A comment about this cast would be nice, explaining how it is assumed that the `placeholders` array can only contain Strings and Futures.